### PR TITLE
Support connecting with a connection string

### PR DESCRIPTION
**Pull request by MattIrv**
This is part of the changes for https://github.com/Microsoft/vscode-mssql/issues/786

- Add a ConnectionString property to ConnectionDetails, and use it to connect in place of the other parameters, if given
- Add a test that attempts to connect using a connection string

### CODE REVIEW

**Comment by MattIrv** [comment]
There are some whitespace changes in the diff because some lines in the files I edited previously had Windows line endings. To view the diff without them, visit https://github.com/Microsoft/sqltoolsservice/pull/334/files?w=1

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/ConnectionServiceTests.cs [hunk 1]
```diff
@@ -1219,15 +1219,5 @@ public void ConnectParamsAreInvalidIfConnectionIsNull()
 
             Assert.NotEmpty(connectionResult.ConnectionId);
         }
-
-        [Fact]
-        public async Task ConnectionWithBadConnectionStringFails()
```
Comment (abist) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/ConnectionServiceTests.cs]: I think this is still a valid test that checks the format of a connection string, isn't it?

Comment (MattIrv) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/ConnectionServiceTests.cs]: That functionality is already tested by BuildConnectionStringWithInvalidAuthType, which is why I removed this test.

File: src/Microsoft.SqlTools.ServiceLayer/Connection/Contracts/ConnectionDetails.cs [hunk 2]
```diff
@@ -485,33 +501,33 @@ private void SetOptionValue<T>(string name, T value)
...
+
+        public bool IsComparableTo(ConnectionDetails other)
+        {
+            if (other == null)
+            {
+                return false;
+            }
+
+            if (!string.Equals(ServerName, other.ServerName)
+                || !string.Equals(AuthenticationType, other.AuthenticationType)
```
Comment (benrr101) [inline src/Microsoft.SqlTools.ServiceLayer/Connection/Contracts/ConnectionDetails.cs]: Generally speaking, this string comparison should be done with `==` unless there's a need for case insensitive comparison.

File: src/Microsoft.SqlTools.ServiceLayer/Connection/Contracts/ConnectionDetails.cs [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 using System;
 using System.Collections.Generic;
-using System.Globalization;
+using System.Globalization;
```
Comment (benrr101) [inline src/Microsoft.SqlTools.ServiceLayer/Connection/Contracts/ConnectionDetails.cs]: I don't see anything that requires Globalization. Let's double check that it's needed before adding.

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/ConnectionServiceTests.cs [hunk 4]
```diff
@@ -1210,5 +1210,14 @@ public void ConnectParamsAreInvalidIfConnectionIsNull()
             await Assert.ThrowsAsync<InvalidOperationException>(
                 () => service.GetOrOpenConnection(TestObjects.ScriptUri, ConnectionType.Query));
         }
+
+        [Fact]
+        public async Task ConnectionWithConnectionStringSucceeds()
```
Comment (benrr101) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/ConnectionServiceTests.cs]: I think since it's possible to have both a connection string and the piece-meal connection properties, we should have a test for that, to make sure the expected behavior is produced.

File: src/Microsoft.SqlTools.ServiceLayer/Connection/Contracts/ConnectionDetails.cs (hunk 3 above)
Comment (MattIrv) [inline src/Microsoft.SqlTools.ServiceLayer/Connection/Contracts/ConnectionDetails.cs]: This is unrelated to my change except that my change restored a unix line ending on this line.

Globalization is used near the bottom of the file by some code that isn't part of my change.

File: src/Microsoft.SqlTools.ServiceLayer/Connection/Contracts/ConnectionDetails.cs (hunk 2 above)
Comment (MattIrv) [inline src/Microsoft.SqlTools.ServiceLayer/Connection/Contracts/ConnectionDetails.cs]: These are unrelated to my changes but were included because the line ending changes. In any case I went ahead and replaced with ==

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/ConnectionServiceTests.cs (hunk 4 above)
Comment (MattIrv) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/ConnectionServiceTests.cs]: added in the new commit

File: src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs [hunk 5]
```diff
@@ -294,7 +294,7 @@ private ConnectionCompleteParams GetConnectionCompleteParams(string connectionTy
                     {
                         ServerName = connectionInfo.ConnectionDetails.ServerName,
                         DatabaseName = connectionInfo.ConnectionDetails.DatabaseName,
-                        UserName = connectionInfo.ConnectionDetails.UserName
+                        UserName = connectionInfo.ConnectionDetails.UserName,
```
Comment (abist) [inline src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs]: extra comma in the end
